src/charm.py

```diff
@@ -0,0 +1,391 @@
+            return
+        if not self._default_database_is_available:
+            self.unit.status = WaitingStatus(
+                "Waiting for `default-database` relation to be available"
+            )
+            return
+        if not self._smf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for `smf-database` relation to be available")
+            return
+        if not self._nrf_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF relation to be available")
+            return
+        if not self._storage_is_attached:
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        if not self._ue_config_file_is_written:
+            event.defer()
+            self.unit.status = WaitingStatus(
+                f"Waiting for `{UEROUTING_CONFIG_FILE}` config file to be pushed to workload container"  # noqa: W505, E501
+            )
+            return
+        content = self._render_config_file(
+            default_database_name=DEFAULT_DATABASE_NAME,
+            default_database_url=self._smf_database_data["uris"].split(",")[0],
+            smf_database_name=SMF_DATABASE_NAME,
+            smf_url=self._smf_hostname,
+            smf_sbi_port=SMF_SBI_PORT,
+            nrf_url=self._nrf_requires.nrf_url,
+            pod_ip=str(self._pod_ip),
+        )
+        self._write_config_file(content=content)
+        self._configure_pebble()
+
+    def _configure_pebble(self) -> None:
+        """Adds layer to pebble config if the proposed config is different from the current one."""
+        self._container.add_layer("smf", self._pebble_layer, combine=True)
+        self._container.replan()
+        self.unit.status = ActiveStatus()
+
+    def _write_config_file(self, content: str) -> None:
+        """Writes config file to workload.
+
+        Args:
+            content (str): Config file content.
+        """
+        self._container.push(
+            path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE}", source=content, make_dirs=True
+        )
+        logger.info("Pushed: %s to workload.", CONFIG_FILE)
+
+    def _write_ue_config_file(self) -> None:
+        """Writes UE config file to workload."""
+        with open(f"src/{UEROUTING_CONFIG_FILE}", "r") as f:
+            content = f.read()
+
+        self._container.push(
+            path=f"{BASE_CONFIG_PATH}/{UEROUTING_CONFIG_FILE}", source=content, make_dirs=True
+        )
+        logger.info("Pushed %s config file to workload", UEROUTING_CONFIG_FILE)
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns whether a given Juju relation was crated.
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            bool: Whether the relation was created.
+        """
+        return bool(self.model.get_relation(relation_name))
+
+    @property
```

Comment:
`_storage_is_attached`, `_config_file_is_written`, `_ue_config_file_is_written`, and honestly most places where we used properties shouldn't be properties. The fact that the storage is attached or not is not an attribute of the charm. It's something that the charm can know by make a call to pebble.

This is probably applicable to all/most properties here. The only exception I see is `_smf_hostname`.

I was guilty of this at a couple of places but I'm trying to remove that kind of design now.

Reply:
Done.